### PR TITLE
Add K-Nearest Neighbors classifier

### LINKED ISSUE #7562: [FEATURE REQUEST] <title> Add K-Nearest Neighbors (KNN) Classification Algorithm
### What would you like to Propose?

I propose adding a **K-Nearest Neighbors (KNN)** implementation to the machinelearning package.

The machinelearning package was recently introduced and currently contains only a few supervised learning algorithms. K-Nearest Neighbors is one of the most fundamental machine learning algorithms and is commonly taught alongside Linear Regression and Naive Bayes in introductory machine learning courses. Adding it would naturally expand the package with another core classification algorithm.

KNN is a good fit for this repository because:

- It is one of the most widely used supervised learning algorithms for classification.
- It can be implemented entirely using standard Java (java.util) without any external dependencies, matching the educational and dependency-free style of this repository.
- The algorithm is straightforward to understand and demonstrates important concepts such as distance metrics, nearest-neighbor search, and majority voting.
- It complements the existing LinearRegression and MultinomialNaiveBayesClassifier implementations already present in the machinelearning package.

### Issue details

**Algorithm / problem statement:** Given a set of labeled training samples and a value of k, classify an unseen sample by computing its Euclidean distance to every training sample, selecting the k nearest neighbors, and assigning the class that receives the majority vote.

- src/main/java/com/thealgorithms/machinelearning/KNearestNeighbors.java

- Store training data through a fit() method.
- Predict the class of a single sample using Euclidean distance.
- Predict class labels for multiple samples.
- Majority-vote classification with deterministic tie-breaking (smaller class label wins).
- Input validation for:
  - k <= 0
  - null or empty datasets
  - inconsistent feature dimensions
  - mismatched feature/label counts
  - prediction before fitting

- src/test/java/com/thealgorithms/machinelearning/KNearestNeighborsTest.java

- Unit tests on well-separated datasets.
- Batch prediction tests.
- Tests for invalid inputs and exception handling.
- Tie-breaking behavior.
- Prediction before fitting.
- Dimension mismatch tests.

- The implementation will include Javadocs describing the algorithm, its time complexity (O(n · d + n log n) in this implementation because neighbors are sorted), and a reference to the Wikipedia article.

### Additional Information

I searched the repository and existing pull requests and could not find a K-Nearest Neighbors implementation in the Java repository. A previous pull request (#2484) titled "KNN Algorithm" was closed as invalid because it did not contain an actual KNN implementation.

If this proposal is accepted, I'd be happy to implement the algorithm along with comprehensive JUnit 5 tests and documentation following the repository's coding standards.

### PR DESCRIPTION
**Pull request by poorva0405**
## Description

This PR adds a K-Nearest Neighbors (KNN) classifier implementation to the `machinelearning` package.

### Features
- Implements KNN classification using Euclidean distance.
- Supports training through `fit()`.
- Supports prediction for single and batch samples.
- Uses deterministic tie-breaking when multiple classes receive the same number of votes.
- Includes comprehensive input validation.
- Includes JUnit 5 tests covering normal and edge cases.
- Includes Javadocs for all public APIs.

Closes #7562

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new algorithms include a corresponding test class that validates their functionality.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/main/java/com/thealgorithms/machinelearning/KNearestNeighbors.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        for (int i = 0; i < trainingFeatures.length; i++) {
+            double distance = euclideanDistance(trainingFeatures[i], testPoint);
+            neighbors.add(new Neighbor(distance, trainingLabels[i]));
+        }
+
+        neighbors.sort(Comparator.comparingDouble(neighbor -> neighbor.distance));
+
+        Map<Integer, Integer> votes = new HashMap<>();
+
+        if (k > trainingFeatures.length) {
```
Comment (AKASH02-byte) [inline src/main/java/com/thealgorithms/machinelearning/KNearestNeighbors.java]: The check if (k > trainingFeatures.length) currently happens inside predict(). Consider moving this validation check into fit() so that invalid $k$ values are caught early during dataset initialization rather than failing later during inference.

**Review by AKASH02-byte — COMMENTED** [review]
Left an inline comment regarding exception checking in KNearestNeighbors.java. Overall, the implementation and tests look clean!

File: src/main/java/com/thealgorithms/machinelearning/KNearestNeighbors.java (hunk 1 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/machinelearning/KNearestNeighbors.java]: +1 to moving this into `fit()`. A few additional points in favor:

1. As written, the check fires only *after* the full distance pass and the `O(n log n)` sort, so all that work is wasted before the exception is thrown.
2. From `predict()` an `IllegalArgumentException` is semantically off — the test point itself is already validated at that point; the real problem is the classifier's configuration vs. the fitted data, which would be an `IllegalStateException`. In `fit()` IAE becomes correct.
3. `throwsExceptionWhenKIsGreaterThanNumberOfTrainingSamples` will need to expect the throw from `fit()` instead:

```java
KNearestNeighbors knn = new KNearestNeighbors(7);
assertThrows(IllegalArgumentException.class, () -> knn.fit(features, labels));
```

**Review by alxkm — CHANGES_REQUESTED** [review]
Solid PR - clean implementation, good tests.

One change requested: move the `k > trainingFeatures.length` check into `fit()`, per the thread on line 168.

Improvements (non-blocking):
- fields are declared mid-class, move them to the top next to `k`
- `fit()` stores the caller's arrays directly — a defensive copy would be safer
- the two "not fitted" null checks in `predict()` can be collapsed into one
- `Neighbor` could be a `record`

Nice tests coverage.

**Review by AKASH02-byte — APPROVED** [review]

**Comment by poorva0405** [comment]
Hi @alxkm, I’ve addressed the requested changes, including moving the k > trainingFeatures.length validation to fit() and implementing the additional improvements discussed.

All 7 CI checks are now passing. Could you please take another look and re-review the PR when you get a chance?

Thank you for the feedback!

**Review by alxkm — APPROVED** [review]
Looks good. Thank you for the contribution.
